src/modules/rover_differential/RoverDifferentialControl/RoverDifferentialControl.cpp

```diff
@@ -0,0 +1,178 @@
+			   1.f, // Integral limit
+			   1.f); // Output limit
+	pid_set_parameters(&_pid_throttle,
+			   _param_rd_p_gain_speed.get(), // Proportional gain
+			   _param_rd_i_gain_speed.get(), // Integral gain
+			   0.f, // Derivative gain
+			   1.f, // Integral limit
+			   1.f); // Output limit
+	pid_set_parameters(&_pid_yaw,
+			   _param_rd_p_gain_yaw.get(),  // Proportional gain
+			   _param_rd_i_gain_yaw.get(),  // Integral gain
+			   0.f,  // Derivative gain
+			   _max_yaw_rate,  // Integral limit
+			   _max_yaw_rate);  // Output limit
+}
+
+void RoverDifferentialControl::computeMotorCommands(const float vehicle_yaw, const float vehicle_yaw_rate,
+		const float vehicle_forward_speed)
+{
+	// Timestamps
+	hrt_abstime timestamp_prev = _timestamp;
+	_timestamp = hrt_absolute_time();
+	const float dt = math::constrain(_timestamp - timestamp_prev, 1_ms, 5000_ms) * 1e-6f;
+
+	// Update differential setpoint
+	_rover_differential_setpoint_sub.update(&_rover_differential_setpoint);
+
+	// Closed loop yaw control (Overrides yaw rate setpoint)
+	if (_rover_differential_setpoint.closed_loop_yaw && PX4_ISFINITE(_rover_differential_setpoint.yaw_setpoint)) {
+		if (fabsf(_rover_differential_setpoint.yaw_setpoint - vehicle_yaw) < FLT_EPSILON) {
+			_rover_differential_setpoint.yaw_rate_setpoint = 0.f;
+			pid_reset_integral(&_pid_yaw);
+
+		} else {
+			const float heading_error = matrix::wrap_pi(_rover_differential_setpoint.yaw_setpoint - vehicle_yaw);
+			_rover_differential_setpoint.yaw_rate_setpoint = pid_calculate(&_pid_yaw, heading_error, 0, 0, dt);
+		}
+	}
+
+	// Yaw rate control
+	float speed_diff_normalized{0.f};
+
+	if (_rover_differential_setpoint.closed_loop_yaw_rate
+	    && PX4_ISFINITE(_rover_differential_setpoint.yaw_rate_setpoint)) { // Closed loop yaw rate control
+		if (fabsf(_rover_differential_setpoint.yaw_rate_setpoint) < FLT_EPSILON) {
+			speed_diff_normalized = 0.f;
+			pid_reset_integral(&_pid_yaw_rate);
+
+		} else {
+			const float speed_diff = _rover_differential_setpoint.yaw_rate_setpoint * _param_rd_wheel_track.get(); // Feedforward
+			speed_diff_normalized = math::interpolate<float>(speed_diff, -_param_rd_max_speed.get(),
+						_param_rd_max_speed.get(), -1.f, 1.f);
+			speed_diff_normalized = math::constrain(speed_diff_normalized +
+								pid_calculate(&_pid_yaw_rate, _rover_differential_setpoint.yaw_rate_setpoint, vehicle_yaw_rate, 0, dt),
+								-1.f, 1.f); // Feedback
+		}
+
+	} else { // Interpret yaw rate setpoint as already normalized
+		speed_diff_normalized = PX4_ISFINITE(_rover_differential_setpoint.yaw_rate_setpoint) ?
+					math::constrain(_rover_differential_setpoint.yaw_rate_setpoint, -1.f, 1.f) : 0.f;
+	}
+
+	// Speed control
+	float throttle{0.f};
+
+	if (_rover_differential_setpoint.closed_loop_speed
+	    && PX4_ISFINITE(_rover_differential_setpoint.forward_speed_setpoint)) { // Closed loop speed control
+		if (fabsf(_rover_differential_setpoint.forward_speed_setpoint) < FLT_EPSILON) {
```

Comment:
What about a negative setpoint, to drive backwards? Is that still possible in a Manual mode?

Reply:
Yes, this check is on the **absolute** value of the speed setpoint.

Reply:
sorry missed that part:)